warehouse/manage/views.py

```diff
@@ -1059,6 +1069,206 @@ def toggle_2fa_requirement(self):
+            )
+
+        return self.default_response
+
+    @view_config(request_method="POST", request_param=GitHubProviderForm.__params__)
+    def add_github_oidc_provider(self):
+        if self.request.flags.enabled(AdminFlagValue.DISALLOW_OIDC):
+            self.request.session.flash(
+                (
+                    "OpenID Connect is temporarily disabled. "
+                    "See https://pypi.org/help#admin-intervention for details."
+                ),
+                queue="error",
+            )
+            return self.default_response
+
+        try:
+            self._check_ratelimits()
+        except TooManyOIDCRegistrations as exc:
+            return HTTPTooManyRequests(
+                self.request._(
+                    "There have been too many attempted OpenID Connect registrations. "
+                    "Try again later."
+                ),
+                retry_after=exc.resets_in.total_seconds(),
+            )
+
+        form = GitHubProviderForm(
+            self.request.POST,
+            api_token=self.request.registry.settings.get("github.token"),
+        )
+
+        if form.validate():
+            self._hit_ratelimits()
+            # GitHub OIDC providers are unique on the tuple of
+            # (repository_name, owner, workflow_name), so we check for
+            # an already registered one before creating.
+            provider = (
+                self.request.db.query(GitHubProvider)
+                .filter(
+                    GitHubProvider.repository_name == form.repository.data,
+                    GitHubProvider.owner == form.normalized_owner,
+                    GitHubProvider.workflow_name == form.workflow_name.data,
+                )
+                .one_or_none()
+            )
+            if provider is None:
+                provider = GitHubProvider(
+                    repository_name=form.repository.data,
+                    owner=form.normalized_owner,
+                    owner_id=form.owner_id,
+                    workflow_name=form.workflow_name.data,
+                )
+
+                self.request.db.add(provider)
+
+            # Each project has a unique registration of OIDC providers,
+            # even though the relationship is many-many.
+            if provider in self.project.oidc_providers:
+                self.request.session.flash(
+                    f"{provider} is already registered with {self.project.name}",
+                    queue="error",
+                )
+                return {**self.default_response, "github_provider_form": form}
+
+            # We only email project owners, since only owners can manage OIDC providers.
+            owner_roles = (
+                self.request.db.query(Role)
+                .filter(Role.project == self.project)
+                .filter(Role.role_name == "Owner")
+                .all()
+            )
+            owner_users = {owner.user for owner in owner_roles}
```

Comment:
This is an N+1 query: for every `Role.project` we do a query for that relationship, and for every `owner.user` we do a query to satisfy that relationship.

We should do this with a single query instead:

```suggestion
            owner_roles = (
                request.db.query(User.id)
                .join(Role.user)
                .filter(Role.role_name == "Owner", Role.project == self.project)
                .subquery()
            )
            owner_users =  (
                request.db.query(User)
                .join(owner_roles, User.id == owner_roles.c.id)
                .all()
            )
```